django_elasticsearch_dsl/documents.py

```diff
@@ -44,6 +48,62 @@
     models.URLField: TextField,
 }
 
+class PagingQuerysetProxy(object):
+    """
+    I am a tiny standin for Django Querysets that implements enough of
+    the protocol (namely count() and __iter__) to be useful for indexing
+    large data sets.
+
+    When iterated over, I will:
+        - use qs.iterator() to disable result set caching in queryset.
+        - chunk fetching the results so that caching in database driver
+            (especially psycopg2) is kept to a minimum, and database
+            drivers that do not support streaming (eg. mysql) do not
+            need to load the whole dataset at once.
+    """
+    def __init__(self, qs, chunk_size=10000):
+        self.qs = qs
+        self.chunk_size = chunk_size
+
+    def count(self):
+        """Pass through to underlying queryset"""
+        return self.qs.count()
+
+    def __iter__(self):
+        """Iterate over result set. Internally uses iterator() as not
+        to cache in the queryset; also supports chunking fetching data
+        in smaller sets so that databases that do not use server side
+        cursors (django docs say only postgres and oracle do) or other
+        optimisations keep memory consumption manageable."""
+
+        last_max_pk = None
+
+        # Get a clone of the QuerySet so that the cache doesn't bloat up
+        # in memory. Useful when reindexing large amounts of data.
+        small_cache_qs = self.qs.order_by('pk')
+
+        once = no_data = False
+        while not no_data and not once:
+            # If we got the max seen PK from last batch, use it to restrict the qs
+            # to values above; this optimises the query for Postgres as not to
+            # devolve into multi-second run time at large offsets.
+            if self.chunk_size:
+                if last_max_pk is not None:
+                    current_qs = small_cache_qs.filter(pk__gt=last_max_pk)[:self.chunk_size]
+                else:
+                    current_qs = small_cache_qs[:self.chunk_size]
```

Comment:
Whats the reason to slice before running the `iterator`? [Django handle it by default if you pass certain ](https://docs.djangoproject.com/en/2.2/ref/models/querysets/) chunk_size`